src/builders/thermodynamic.jl

```diff
@@ -0,0 +1,258 @@
+"""
+$(TYPEDSIGNATURES)
+
+The function uses the supplied `optimizer` and
+`reaction_standard_gibbs_free_energies`. Optionally, `flux_solution` can be used
+to set the directions of each reaction in `model` (all reactions are assumed to
+proceed forward and are active by default). The supplied `flux_solution` should
+be free of internal cycles i.e. thermodynamically consistent. This optional
+input is important if a reaction in `model` normally runs in reverse (negative
+flux). Note, reactions in `flux_solution` that are smaller than `small_flux_tol`
+are also ignored in the analysis function (for numerical stability).
+
+
+Reactions specified in `ignore_reaction_ids` are internally ignored when
+calculating the max-min driving force. This should include water and proton
+importers.
+
+Since biochemical thermodynamics are assumed, the `proton_ids` and `water_ids`
+need to be specified so that they can be ignored in the calculations.
+Effectively this assumes an aqueous environment at constant pH is used.
+
+`constant_concentrations` is used to fix the concentrations of certain
+metabolites (such as CO₂). `concentration_ratios` is used to specify additional
+constraints on metabolite pair concentrations (typically, this is done with
+various cofactors such as the ATP/ADP ratio. For example, you can fix the
+concentration of ATP to be always 5× higher than of ADP by specifying
+`Dict(("ATP","ADP") => 5.0)`
+
+`concentration_lb` and `concentration_ub` set default concentration bounds, in M
+by default.
+
+`T` and `R` can be specified in the corresponding units; defaults are K and
+kJ/K/mol.
+"""
+function build_max_min_driving_force_model(
+    model::A.AbstractFBCModel,
+    reaction_standard_gibbs_free_energies::Dict{String,Float64};
+    flux_solution = Dict{String,Float64}(),
+    proton_ids = ["h_c", "h_e"],
+    water_ids = ["h2o_c", "h2o_e"],
+    concentration_lb = 1e-9,
+    concentration_ub = 100e-3,
+    T = 298.15, # Kelvin
+    R = 8.31446261815324e-3, # kJ/K/mol
+    ignore_reaction_ids = String[],
+)
+    # check if reactions that will be used in the model all have thermodynamic data, otherwise throw error.
+    # only these reactions will form part of the model.
+    rxns = filter(
+        x -> !(x in ignore_reaction_ids),
+        isempty(flux_solution) ? A.reactions(model) : collect(keys(flux_solution)),
+    )
+    all(in.(rxns, Ref(collect(keys(reaction_standard_gibbs_free_energies))))) || throw(
+        ArgumentError("""
```

Comment:
this is DomainError

Reply:
...on a side thought, why not just consider the reactions that have thermo data instead of forcing people to make sure that every reaction that doesn't have the thermo data gets filled precisely into the ignored ids?

Reply:
To prevent silent footguns. This way is more explicit. We had it the other way around before, but I don't like the silence